### PR TITLE
Enhance outgoing webhooks and add history

### LINKED ISSUE #5176: Outgoing webhook trigger on file upload
Your Rocket.Chat version: 0.47

I am creating a custom integration with rocket.chat. I want to use the outgoing webhook triggered at file upload. Is this something that is possible with the current release or can be made available in the future?

### LINKED ISSUE #5226: Webhook for new channel created
Your Rocket.Chat version: 0.46

I need to know when a new channel is created but the outgoing webhooks feature seems to just handle watching of messages. (like Slack do)

If there is no other way to get an event when a new channel is created I'll add a webhook triggered by callback "afterCreateChannel" to handle it.

Edit: The webhooks/integrations feature is totally centered on messages events. Hard to believe I can just "add" channels events.

Please @RocketChat/core let me know if you think that webhooks for channels related events is a good/bad idea.

### PR DESCRIPTION
**Pull request by graywolf336**
@RocketChat/core 

Closes #5226
Closed #5176 

## Integrations List
image

## New Integration
image

## New Outgoing Integration
image

## Outgoing Integration Overview
image

## Outgoing Integration Advanced Settings
image

## Outgoing Integration History List
image

## Outgoing Integration History List Item
image

### CODE REVIEW

File: packages/rocketchat-i18n/i18n/en.i18n.json [hunk 1]
```diff
@@ -637,6 +640,9 @@
   "Integration_added": "Integration has been added",
   "Integration_Incoming_WebHook": "Incoming WebHook Integration",
   "Integration_New": "New Integration",
+  "Integrations_Outgoing_Type_RoomCreated": "Room Created (public and private)",
+  "Integrations_Outgoing_Type_SendMessage": "Message Sent",
+	"Integrations_Outgoing_Type_UserCreated": "User Created",
```
Comment (graywolf336) [inline packages/rocketchat-i18n/i18n/en.i18n.json]: Bah! This file

**Comment by graywolf336** [comment]
@engelgabriel @marceloschmidt @rodrigok this is now ready to be reviewed :+1:

File: packages/rocketchat-integrations/client/views/integrationsNew.html [hunk 2]
```diff
@@ -1,44 +1,52 @@
...
-									</div>
-									<div class="admin-integrations-new-item-description">
-										Get data out of Rocket.Chat in real-time.
+							</a>
+							<a href="{{pathFor "admin-integrations-outgoing"}}">
+								<div class="admin-integrations-new-item">
+									<i class="icon-logout"></i>
+									<div class="admin-integrations-new-item-body">
+										<div class="admin-integrations-new-item-title">
+											Outgoing WebHook
```
Comment (marceloschmidt) [inline packages/rocketchat-integrations/client/views/integrationsNew.html]: Missing translation

File: packages/rocketchat-integrations/client/views/integrationsNew.html [hunk 3]
```diff
@@ -1,44 +1,52 @@
...
+										</div>
+										<div class="admin-integrations-new-item-description">
+											Get data out of Rocket.Chat in real-time.
```
Comment (marceloschmidt) [inline packages/rocketchat-integrations/client/views/integrationsNew.html]: Missing translation

File: packages/rocketchat-integrations/lib/rocketchat.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
+RocketChat.integrations = {
```
Comment (marceloschmidt) [inline packages/rocketchat-integrations/lib/rocketchat.js]: Don't we usually capitalize first letter on those? Eg. `RocketChat.Cache`, `RocketChat.SlackBridge`, `RocketChat.Settings`

Although there's also an example for `RocketChat.models`. Just curious

File: packages/rocketchat-integrations/server/lib/validation.js [hunk 5]
```diff
@@ -0,0 +1,155 @@
+/* global Babel */
+const scopedChannels = ['all_public_channels', 'all_private_groups', 'all_direct_messages'];
+const validChannelChars = ['@', '#'];
+
+function _verifyRequiredFields(integration) {
+	if (!integration.event || !Match.test(integration.event, String) || integration.event.trim() === '' || !RocketChat.integrations.outgoingEvents[integration.event]) {
+		throw new Meteor.Error('error-invalid-event-type', 'Invalid event type', { function: 'validateOutgoing' });
```
Comment (marceloschmidt) [inline packages/rocketchat-integrations/server/lib/validation.js]: Shouldn't the `function` here say `_verifyRequiredFields`?

File: packages/rocketchat-integrations/server/lib/validation.js [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+	}
+
+	if (!integration.username || !Match.test(integration.username, String) || integration.username.trim() === '') {
+		throw new Meteor.Error('error-invalid-username', 'Invalid username', { function: 'validateOutgoing' });
```
Comment (marceloschmidt) [inline packages/rocketchat-integrations/server/lib/validation.js]: Likewise above and below.

File: packages/rocketchat-integrations/server/lib/validation.js [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+	}
+}
+
+function _verifyUserHasPermissionForChannels(integration, userId, channels) {
+	for (let channel of channels) {
+		if (scopedChannels.includes(channel)) {
+			if (channel === 'all_public_channels') {
+				// No special permissions needed to add integration to public channels
+			} else if (!RocketChat.authz.hasPermission(userId, 'manage-integrations')) {
+				throw new Meteor.Error('error-invalid-channel', 'Invalid Channel', { function: 'validateOutgoing' });
```
Comment (marceloschmidt) [inline packages/rocketchat-integrations/server/lib/validation.js]: Function name again

File: packages/rocketchat-lib/lib/callbacks.coffee [hunk 8]
```diff
@@ -92,7 +92,7 @@ RocketChat.callbacks.run = (hook, item, constant) ->
 					else
 						console.log String(currentTime), hook, callback.id, callback.stack?.split?('\n')[2]?.match(/\(.+\)/)?[0]
 
-			return callbackResult
+			return if typeof callbackResult == 'undefined' then result else callbackResult
```
Comment (marceloschmidt) [inline packages/rocketchat-lib/lib/callbacks.coffee]: Should use `===` instead of `==`

Comment (graywolf336) [inline packages/rocketchat-lib/lib/callbacks.coffee]: Good catch.

File: packages/rocketchat-integrations/server/lib/validation.js (hunk 5 above)
Comment (graywolf336) [inline packages/rocketchat-integrations/server/lib/validation.js]: Honestly not sure, as these are internal ones and the "bigger" function that is being used is actually `validateOutgoing`.

File: packages/rocketchat-integrations/lib/rocketchat.js (hunk 4 above)
Comment (graywolf336) [inline packages/rocketchat-integrations/lib/rocketchat.js]: Yeah, we don't have anything consistent. This is just what the coffeescript had and was already being referenced by, so I kept it with that.

File: packages/rocketchat-lib/lib/callbacks.coffee (hunk 8 above)
Comment (graywolf336) [inline packages/rocketchat-lib/lib/callbacks.coffee]: Actually no it shouldn't, as this file is coffeescript. ;)

**Comment by graywolf336** [comment]
@rodrigok Can I get this reviewed so we can start testing it?

**Pull request merged**